internal/crypto/cms/signed.go

```diff
@@ -0,0 +1,216 @@
+package cms
+
+import (
+	"bytes"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"time"
+
+	"github.com/notaryproject/notation-go-lib/internal/crypto/hashutil"
+	"github.com/notaryproject/notation-go-lib/internal/crypto/oid"
+)
+
+// ParsedSignedData is a parsed SignedData structure for golang friendly types.
+type ParsedSignedData struct {
+	Content      []byte
+	ContentType  asn1.ObjectIdentifier
+	Certificates []*x509.Certificate
+	CRLs         []pkix.CertificateList
+	Signers      []SignerInfo
+}
+
+// ParseSignedData parses ASN.1 DER-encoded SignedData structure to golang friendly types.
+func ParseSignedData(data []byte) (*ParsedSignedData, error) {
+	var contentInfo ContentInfo
+	if _, err := asn1.Unmarshal(data, &contentInfo); err != nil {
+		return nil, SyntaxError{Message: "invalid content info", Detail: err}
+	}
+	if !oid.SignedData.Equal(contentInfo.ContentType) {
+		return nil, ErrExpectSignedData
+	}
+
+	var signedData SignedData
+	if _, err := asn1.Unmarshal(contentInfo.Content.Bytes, &signedData); err != nil {
+		return nil, SyntaxError{Message: "invalid signed data", Detail: err}
+	}
+	certs, err := x509.ParseCertificates(signedData.Certificates.Bytes)
+	if err != nil {
+		return nil, SyntaxError{Message: "invalid signed data", Detail: err}
+	}
+
+	return &ParsedSignedData{
+		Content:      signedData.EncapsulatedContentInfo.Content,
+		ContentType:  signedData.EncapsulatedContentInfo.ContentType,
+		Certificates: certs,
+		CRLs:         signedData.CRLs,
+		Signers:      signedData.SignerInfos,
+	}, nil
+}
+
+// Verify attempts to verify the content in the parsed signed data against the signer
+// information. The `Intermediates` in the verify options will be ignored and
+// re-contrusted using the certificates in the parsed signed data.
+// If more than one signature is present, the successful validation of any signature
+// implies that the content in the parsed signed data is valid.
+// On successful verification, the list of signers that successfully verify is returned.
+// If all signatures fail to verify, the last error is returned.
+// References:
+// - RFC 5652 5   Signed-data Content Type
+// - RFC 5652 5.4 Message Digest Calculation Process
+// - RFC 5652 5.6 Signature Verification Process
+// WARNING: this function doesn't do any revocation checking.
+func (d *ParsedSignedData) Verify(opts x509.VerifyOptions) ([]SignerInfo, error) {
+	if len(d.Signers) == 0 {
+		return nil, ErrSignerNotFound
+	}
+	if len(d.Certificates) == 0 {
+		return nil, ErrCertificateNotFound
+	}
+
+	intermediates := x509.NewCertPool()
```

Comment:
nit: maybe you can add a comment to highlight this replacement part

Reply:
This replacement part has been mentioned in the documentation of this method. See line 52.